Add link validation (fix #122)

lukpueh opened the pull request:
This PR adds validation methods to `link.LinkSignable` that check the format of the corresponding fields. (closes #122)

By calling `self.validate()` on a`LinkSignable` object  all fields' formats are checked.

This PR also
- adds validation to `LinkSignable`'s constructor,
- adapts existing unit tests where schema validation was less strict, and
- adds new unit tests for the new validation methods.

Commit:
Add schema validation methods for LinkSignable

This commit adds validation methods to LinkSignable that check
the format of the fields of LinkSignable objects.
By calling `self.validate()`, inherited from
`common.ValidationMixin`, on a LinkSignable object the entire
object is checked.

This commit also adds a call to validate to the LinkSignable's
constructor.

Commit:
Adapt unit tests to LinkSignable validation

This commit fixes wrong formats in a unit test and
in a demo link file used by many unit tests.

Commit:
Add unit tests for link validators

SantiagoTorres commented on in_toto/models/link.py:
```diff
@@ -141,3 +141,59 @@ def __init__(self, **kwargs):
     self.return_value = kwargs.get("return_value", None)
+
+    self.validate()
+
```
I think you need another space here, or less space on the ones below

SantiagoTorres commented on in_toto/models/link.py:
```diff
@@ -141,3 +141,59 @@ def __init__(self, **kwargs):
+    """Private method to check that `materials` is a `dict` of `HASHDICTs`."""
+    if not isinstance(self.materials, dict):
+      raise securesystemslib.exceptions.FormatError(
+          "Invalid Link: field `materials` must be of type list, got: {}"
```
I wonder if we can coalesce this to a format string (probably with our own exception class) and just do:
```
raise InvalidLinkException(field='materials', got=type(self.materials))
```

Here and elsewhere.

lukpueh replied:
We could and should, but let's do this elsewhere. Care for a "revisit exception taxonomy" ticket?

lukpueh replied:
FYI: https://github.com/in-toto/in-toto/issues/126

SantiagoTorres commented on test/models/test_link.py:
```diff
@@ -0,0 +1,124 @@
+  Lukas Puehringer <lukas.puehringer@nyu.edu>
+
+<Started>
+  Sep 28, 2011
```
I don't think this is the right date

SantiagoTorres commented on test/models/test_link.py:
```diff
@@ -0,0 +1,124 @@
+    test_link = Link()
+
+    # Good return_value 1
+    test_link.signed.retunr_value = 1
```
I don't think this is checking what it should check.

SantiagoTorres reviewed (changes requested):
Four really minor details. I think once these are addressed this is ready to merge.

Commit:
Make spacing consistent in link.py

- 1 blank line inside functions
- 2 blank lines between functions
- 3 blank lines between classes

We should adapt the rest of the code and also discuss
this in our code style guidelines. The current suggestion (1/2/5)
is too much, imho.

Commit:
Fix bug and docs/comment in test_link

lukpueh commented:
Thanks for the comments, @SantiagoTorres. Please merge when ready!
